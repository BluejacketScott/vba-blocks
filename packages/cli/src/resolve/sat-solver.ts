import { satisfies } from 'semver';
import { Solver, exactlyOne, atMostOne, implies, or } from 'logic-solver';

import { Config } from '../config';
import { Manifest, Dependency } from '../manifest';
import {
  isRegistryDependency,
  isPathDependency,
  isGitDependency
} from '../manifest/dependency';
import { Registration, getRegistrationId } from '../manager';
import { DependencyGraph } from './dependency-graph';
import Resolver, { Resolution, ResolutionGraph } from './resolver';
import { unique } from '../utils';

export default async function solve(
  config: Config,
  manifest: Manifest,
  resolver: Resolver
): Promise<DependencyGraph> {
  await resolveDependencies(manifest.dependencies, resolver);
  await optimizeResolved(resolver.graph);

  const solver = new Solver();
  const required = manifest.dependencies.map(dependency => dependency.name);

  for (const [name, resolved] of resolver) {
    const { registered } = resolved;
    const isRequired = required.includes(name);
    const ids = registered.map(getRegistrationId);

    if (isRequired) {
      solver.require(exactlyOne(...ids));
    } else {
      solver.require(atMostOne(...ids));
    }
  }

  const reversed = [...resolver].reverse();
  for (const [name, resolved] of reversed) {
    const { registered } = resolved;

    for (const registration of registered) {
      const id = getRegistrationId(registration);

      for (const dependency of registration.dependencies) {
        const resolved = resolver.graph.get(dependency.name);
        const matching = getMatching(dependency, resolved);

        solver.require(implies(id, or(...matching)));
      }
    }
  }

  const solution = solver.solve();

  if (!solution) {
    throw new Error('Failed to resolve dependency graph for given manifest');
  }

  const ids = solution.getTrueVars();
  const graph = ids.map(id => resolver.getRegistration(id));

  return graph;
}

export async function resolveDependencies(
  dependencies: Dependency[],
  resolver: Resolver
): Promise<void> {
  const resolved = await Promise.all(
    dependencies.map(dependency => resolver.get(dependency))
  );

  for (const resolution of resolved) {
    const { registered } = resolution;
    await Promise.all(
      registered.map(registration =>
        resolveDependencies(registration.dependencies, resolver)
      )
    );
  }
}

export async function optimizeResolved(graph: ResolutionGraph): Promise<void> {
  for (const [name, resolution] of graph) {
    const { registered } = resolution;
    const range = unique(resolution.range).join(' || ');

    resolution.registered = registered
      .filter(registration => satisfies(registration.version, range))
      .reverse();
  }
}

function getMatching(dependency: Dependency, resolved: Resolution): string[] {
  const { registered } = resolved;

  if (isRegistryDependency(dependency)) {
    const { version } = dependency;

    return registered
      .filter(registration => satisfies(registration.version, version))
      .map(getRegistrationId);
  } else if (isPathDependency(dependency)) {
    return [];
  } else {
    return [];
  }
}
