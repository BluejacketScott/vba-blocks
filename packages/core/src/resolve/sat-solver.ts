import * as semver from 'semver';
import { Config } from '../config';
import { Manifest, Dependency } from '../manifest';
import { Registration } from '../registry';
import { DependencyGraph } from './dependency-graph';
import Resolver from './resolver';
import * as Logic from 'logic-solver';

export default async function solve(
  config: Config,
  manifest: Manifest,
  resolver: Resolver
): Promise<DependencyGraph> {
  await resolveDependencies(manifest.dependencies, resolver);

  const solver = new Logic.Solver();
  const required = manifest.dependencies.map(dependency => dependency.name);

  for (const resolved of resolver) {
    const { name, registered } = resolved;
    const isRequired = required.includes(name);
    const ids = registered.map(
      registration => `${name}@${registration.version}`
    );

    if (isRequired) {
      solver.require(Logic.exactlyOne(...ids));
    } else {
      solver.require(Logic.atMostOne(...ids));
    }
  }

  const reversed = [...resolver].reverse();
  for (const resolved of reversed) {
    const { name, registered } = resolved;

    for (const registration of registered) {
      const id = `${name}@${registration.version}`;

      for (const dependency of registration.dependencies) {
        const matching = getMatching(dependency, resolver);
        solver.require(Logic.implies(id, Logic.or(...matching)));
      }
    }
  }

  const solution = solver.solve();

  if (!solution) {
    throw new Error('Failed to resolve dependency graph for given manifest');
  }

  const ids = solution.getTrueVars();
  const graph = ids.map(id => {
    const [name, version] = id.split('@');
    const { registered } = resolver.graph.get(name);
    return registered.find(registration => registration.version === version);
  });

  return graph;
}

export async function resolveDependencies(
  dependencies: Dependency[],
  resolver: Resolver
): Promise<void> {
  const resolved = await Promise.all(
    dependencies.map(dependency => resolver.get(dependency))
  );

  for (const resolution of resolved) {
    const { registered } = resolution;
    await Promise.all(
      registered.map(registration =>
        resolveDependencies(registration.dependencies, resolver)
      )
    );
  }
}

function getMatching(dependency: Dependency, resolver: Resolver): string[] {
  const { name, version } = dependency;
  const { registered } = resolver.graph.get(name);

  const matching = registered
    .filter(registration => semver.satisfies(version, registration.version))
    .map(registration => `${name}@${registration.version}`);

  return matching;
}
