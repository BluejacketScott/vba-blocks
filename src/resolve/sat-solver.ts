import { ok } from 'assert';
import * as logicSolver from 'logic-solver';
import { satisfies } from 'semver';
import { Dependency, isRegistryDependency } from '../manifest/dependency';
import { Workspace } from '../professional/workspace';
import { Registration } from '../sources/registration';
import unique from '../utils/unique';
import { DependencyGraph } from './dependency-graph';
import Resolver, { Resolution, ResolutionGraph } from './resolver';

const { Solver, exactlyOne, atMostOne, implies, or } = logicSolver;

export async function solve(workspace: Workspace, resolver: Resolver): Promise<DependencyGraph> {
  const dependencies = workspace.root.dependencies.concat(workspace.root.devDependencies);

  await resolveDependencies(dependencies, resolver);
  const required = await optimizeResolved(resolver.graph, dependencies);

  const solver = new Solver();

  for (const [name, resolved] of resolver) {
    const { registered } = resolved;
    const isRequired = required.includes(name);
    const ids = registered.map(registration => registration.id);

    if (isRequired) {
      solver.require(exactlyOne(...ids));
    } else {
      solver.require(atMostOne(...ids));
    }
  }

  const reversed = [...resolver].reverse();
  for (const [name, resolved] of reversed) {
    const { registered } = resolved;

    for (const registration of registered) {
      for (const dependency of registration.dependencies) {
        const resolved = resolver.graph.get(dependency.name);
        ok(resolved, `Dependency "${dependency.name}" hasn't been resolved`);

        const matching = getMatching(dependency, resolved!);

        solver.require(implies(registration.id, or(...matching)));
      }
    }
  }

  const solution = solver.solve();

  if (!solution) {
    throw new Error('Failed to resolve dependency graph for given manifest');
  }

  const ids: string[] = solution.getTrueVars();

  const graph = ids.map(id => resolver.getRegistration(id)).filter(isRegistration);

  return graph;

  function isRegistration(value: Registration | undefined): value is Registration {
    return value != null;
  }
}

export async function resolveDependencies(
  dependencies: Dependency[],
  resolver: Resolver
): Promise<void> {
  const resolved = await Promise.all(dependencies.map(dependency => resolver.get(dependency)));

  for (const resolution of resolved) {
    const { registered } = resolution;
    await Promise.all(
      registered.map(registration => resolveDependencies(registration.dependencies, resolver))
    );
  }
}

export async function optimizeResolved(
  graph: ResolutionGraph,
  dependencies: Dependency[]
): Promise<string[]> {
  const required: string[] = [];
  const topLevel: { [name: string]: string } = {};

  for (const dependency of dependencies) {
    const { name } = dependency;
    required.push(name);

    if (isRegistryDependency(dependency)) {
      topLevel[name] = dependency.version;
    }
  }

  for (const [name, resolution] of graph) {
    const { registered } = resolution;
    const range = topLevel[name] || unique(resolution.range).join(' || ');

    resolution.registered = registered
      .filter(registration => satisfies(registration.version, range))
      .reverse();
  }

  return required;
}

function getMatching(dependency: Dependency, resolved: Resolution): string[] {
  const { registered } = resolved;

  if (isRegistryDependency(dependency)) {
    const { version } = dependency;

    return registered
      .filter(registration => satisfies(registration.version, version!))
      .map(registration => registration.id);
  } else {
    return [];
  }
}
