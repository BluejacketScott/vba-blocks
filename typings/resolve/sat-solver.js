import { ok } from 'assert';
import { satisfies } from 'semver';
import * as logicSolver from 'logic-solver';
import { isRegistryDependency } from '../manifest/dependency';
import unique from '../utils/unique';
const { Solver, exactlyOne, atMostOne, implies, or } = logicSolver;
export async function solve(workspace, resolver) {
    const dependencies = workspace.root.dependencies;
    await resolveDependencies(dependencies, resolver);
    const required = await optimizeResolved(resolver.graph, dependencies);
    const solver = new Solver();
    for (const [name, resolved] of resolver) {
        const { registered } = resolved;
        const isRequired = required.includes(name);
        const ids = registered.map(registration => registration.id);
        if (isRequired) {
            solver.require(exactlyOne(...ids));
        }
        else {
            solver.require(atMostOne(...ids));
        }
    }
    const reversed = [...resolver].reverse();
    for (const [name, resolved] of reversed) {
        const { registered } = resolved;
        for (const registration of registered) {
            for (const dependency of registration.dependencies) {
                const resolved = resolver.graph.get(dependency.name);
                ok(resolved, `Dependency "${dependency.name}" hasn't been resolved`);
                const matching = getMatching(dependency, resolved);
                solver.require(implies(registration.id, or(...matching)));
            }
        }
    }
    const solution = solver.solve();
    if (!solution) {
        throw new Error('Failed to resolve dependency graph for given manifest');
    }
    const ids = solution.getTrueVars();
    const graph = ids.map(id => resolver.getRegistration(id)).filter(isRegistration);
    return graph;
    function isRegistration(value) {
        return value != null;
    }
}
export async function resolveDependencies(dependencies, resolver) {
    const resolved = await Promise.all(dependencies.map(dependency => resolver.get(dependency)));
    for (const resolution of resolved) {
        const { registered } = resolution;
        await Promise.all(registered.map(registration => resolveDependencies(registration.dependencies, resolver)));
    }
}
export async function optimizeResolved(graph, dependencies) {
    const required = [];
    const topLevel = {};
    for (const dependency of dependencies) {
        const { name } = dependency;
        required.push(name);
        if (isRegistryDependency(dependency)) {
            topLevel[name] = dependency.version;
        }
    }
    for (const [name, resolution] of graph) {
        const { registered } = resolution;
        const range = topLevel[name] || unique(resolution.range).join(' || ');
        resolution.registered = registered
            .filter(registration => satisfies(registration.version, range))
            .reverse();
    }
    return required;
}
function getMatching(dependency, resolved) {
    const { registered } = resolved;
    if (isRegistryDependency(dependency)) {
        const { version } = dependency;
        return registered
            .filter(registration => satisfies(registration.version, version))
            .map(registration => registration.id);
    }
    else {
        return [];
    }
}
