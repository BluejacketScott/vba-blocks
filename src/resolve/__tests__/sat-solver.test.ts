import { solve } from '../sat-solver';
import Resolver from '../resolver';
import { getConfig } from '@vba-blocks/helpers';
import * as workspaces from '@vba-blocks/fixtures/workspaces';

test('solves simple tree', async () => {
  const config = getConfig();
  const resolver = new Resolver(config);

  const solution = await solve(workspaces.simple, resolver);

  expect(solution).toMatchSnapshot();
});

test('solves complex tree', async () => {
  const config = getConfig();
  const resolver = new Resolver(config);

  const solution = await solve(workspaces.complex, resolver);

  expect(solution).toMatchSnapshot();
});

test('solves needs-sat tree', async () => {
  const config = getConfig();
  const resolver = new Resolver(config);

  const solution = await solve(workspaces.needsSat, resolver);

  expect(solution).toMatchSnapshot();
});

test('fails to solve unresolvable tree', async () => {
  const config = getConfig();
  const resolver = new Resolver(config);

  await expect(
    solve(workspaces.unresolvable, resolver)
  ).rejects.toMatchSnapshot();
});
