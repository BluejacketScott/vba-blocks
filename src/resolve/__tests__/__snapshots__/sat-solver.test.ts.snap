// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fails to solve unresolvable tree 1`] = `[Error: Failed to resolve dependency graph for given manifest]`;

exports[`solves complex tree 1`] = `
Array [
  Object {
    "dependencies": Array [],
    "id": "dictionary@1.4.1",
    "name": "dictionary",
    "source": "registry+vba-blocks#sha256-b241062bcd7ffd68c9063e3980a5df4c6781df29d96763a020d291e52454f6a5",
    "version": "1.4.1",
  },
  Object {
    "dependencies": Array [
      Object {
        "name": "dictionary",
        "registry": "vba-blocks",
        "version": "^1",
      },
    ],
    "id": "json@2.3.0",
    "name": "json",
    "source": "registry+vba-blocks#sha256-3c2760ce2940dd19c16542cead792e0d05e58566155b1fbc72d445fe50a36799",
    "version": "2.3.0",
  },
  Object {
    "dependencies": Array [
      Object {
        "name": "json",
        "registry": "vba-blocks",
        "version": "^2.3.0",
      },
    ],
    "id": "web@4.1.5",
    "name": "web",
    "source": "registry+vba-blocks#sha256-56c0bfaa84be448b39c65aa83f1c7530db0635a6b2b6ee8589a18feea29e37e9",
    "version": "4.1.5",
  },
]
`;

exports[`solves needs-sat tree 1`] = `
Array [
  Object {
    "dependencies": Array [],
    "id": "dictionary@1.4.1",
    "name": "dictionary",
    "source": "registry+vba-blocks#sha256-b241062bcd7ffd68c9063e3980a5df4c6781df29d96763a020d291e52454f6a5",
    "version": "1.4.1",
  },
  Object {
    "dependencies": Array [
      Object {
        "name": "dictionary",
        "registry": "vba-blocks",
        "version": "^1",
      },
    ],
    "id": "json@2.1.0",
    "name": "json",
    "source": "registry+vba-blocks#sha256-c00e834ea96c4a6968d167b6616b2fa44a13fb672ff690b60a13326ca18de711",
    "version": "2.1.0",
  },
  Object {
    "dependencies": Array [
      Object {
        "name": "json",
        "registry": "vba-blocks",
        "version": "^2",
      },
    ],
    "id": "web@4.0.22",
    "name": "web",
    "source": "registry+vba-blocks#sha256-253052084974af810a115b031e782b4b3a23fdf55d77594ee5f9e2d76a2d396d",
    "version": "4.0.22",
  },
]
`;

exports[`solves standard tree 1`] = `
Array [
  Object {
    "dependencies": Array [],
    "id": "dictionary@1.4.1",
    "name": "dictionary",
    "source": "registry+vba-blocks#sha256-b241062bcd7ffd68c9063e3980a5df4c6781df29d96763a020d291e52454f6a5",
    "version": "1.4.1",
  },
  Object {
    "dependencies": Array [
      Object {
        "name": "dictionary",
        "registry": "vba-blocks",
        "version": "^1",
      },
    ],
    "id": "json@2.3.0",
    "name": "json",
    "source": "registry+vba-blocks#sha256-3c2760ce2940dd19c16542cead792e0d05e58566155b1fbc72d445fe50a36799",
    "version": "2.3.0",
  },
]
`;
